File: app/src/main/java/com/github/android/lvrn/lvrnproject/persistent/repository/core/impl/TaskRepositoryImpl.java
package com.github.android.lvrn.lvrnproject.persistent.repository.core.impl;

import android.content.ContentValues;
import android.database.Cursor;
import androidx.annotation.NonNull;

import com.github.valhallalabs.laverna.persistent.entity.Task;
import com.github.android.lvrn.lvrnproject.persistent.repository.core.TaskRepository;
import com.github.android.lvrn.lvrnproject.persistent.repository.impl.ProfileDependedRepositoryImpl;
import com.github.android.lvrn.lvrnproject.util.PaginationArgs;

import java.util.List;

import static com.github.android.lvrn.lvrnproject.persistent.database.LavernaContract.TasksTable.COLUMN_DESCRIPTION;
import static com.github.android.lvrn.lvrnproject.persistent.database.LavernaContract.TasksTable.COLUMN_ID;
import static com.github.android.lvrn.lvrnproject.persistent.database.LavernaContract.TasksTable.COLUMN_IS_COMPLETED;
import static com.github.android.lvrn.lvrnproject.persistent.database.LavernaContract.TasksTable.COLUMN_NOTE_ID;
import static com.github.android.lvrn.lvrnproject.persistent.database.LavernaContract.TasksTable.COLUMN_PROFILE_ID;
import static com.github.android.lvrn.lvrnproject.persistent.database.LavernaContract.TasksTable.TABLE_NAME;

/**
 * @author Vadim Boitsov <vadimboitsov1@gmail.com>
 */

public class TaskRepositoryImpl extends ProfileDependedRepositoryImpl<Task> implements TaskRepository {

    public TaskRepositoryImpl() {
        super(TABLE_NAME);
    }

    @NonNull
    @Override
    protected ContentValues toContentValues(@NonNull Task entity) {
        int completed = entity.isCompleted() ? 1 : 0;
        ContentValues contentValues = new ContentValues();
        contentValues.put(COLUMN_ID, entity.getId());
        contentValues.put(COLUMN_PROFILE_ID, entity.getProfileId());
        contentValues.put(COLUMN_NOTE_ID, entity.getNoteId());
        contentValues.put(COLUMN_DESCRIPTION, entity.getDescription());
        contentValues.put(COLUMN_IS_COMPLETED, completed);
        return contentValues;
    }

    @NonNull
    @Override
    protected Task toEntity(@NonNull Cursor cursor) {
        return new Task(
                cursor.getString(cursor.getColumnIndex(COLUMN_ID)),
                cursor.getString(cursor.getColumnIndex(COLUMN_PROFILE_ID)),
                cursor.getString(cursor.getColumnIndex(COLUMN_NOTE_ID)),
                cursor.getString(cursor.getColumnIndex(COLUMN_DESCRIPTION)),
                cursor.getInt(cursor.getColumnIndex(COLUMN_IS_COMPLETED)) > 0);
    }

    @NonNull
    @Override
    public List<Task> getUncompletedByProfile(@NonNull String profileId, @NonNull PaginationArgs paginationArgs) {
        String additionalClause = " AND " + COLUMN_IS_COMPLETED + " = 0";
        return super.getByIdCondition(COLUMN_PROFILE_ID, profileId, additionalClause, paginationArgs);
    }

    @NonNull
    @Override
    public List<Task> getUncompletedByProfileAndDescription(@NonNull String profileId, @NonNull String description, @NonNull PaginationArgs paginationArgs) {
        String additionalClause = " AND " + COLUMN_IS_COMPLETED + " = 0";
        return super.getByName(COLUMN_DESCRIPTION, profileId, description, additionalClause, paginationArgs);
    }

    @NonNull
    @Override
    public List<Task> getByNote(@NonNull String noteId) {
        String query = "SELECT * FROM " + TABLE_NAME
                + " WHERE " + COLUMN_NOTE_ID + " = '" + noteId + "'";
        return getByRawQuery(query);
    }

    @Override
    public boolean update(@NonNull Task entity) {
        int completed = entity.isCompleted() ? 1 : 0;
        String query = "UPDATE " + TABLE_NAME
                + " SET "
                + COLUMN_DESCRIPTION + "='" + entity.getDescription() + "', "
                + COLUMN_IS_COMPLETED + "='" + completed + "' "
                + " WHERE " + COLUMN_ID + "='" + entity.getId() + "'";
        return super.rawUpdateQuery(query);
    }
}

File: app/src/main/java/com/github/android/lvrn/lvrnproject/persistent/repository/core/impl/TaskRepositoryImpl.kt
package com.github.android.lvrn.lvrnproject.persistent.repository.core.impl

import android.content.ContentValues
import android.database.Cursor
import com.github.android.lvrn.lvrnproject.persistent.database.LavernaContract.LavernaBaseTable
import com.github.android.lvrn.lvrnproject.persistent.database.LavernaContract.ProfileDependedTable
import com.github.android.lvrn.lvrnproject.persistent.database.LavernaContract.TasksTable.COLUMN_DESCRIPTION
import com.github.android.lvrn.lvrnproject.persistent.database.LavernaContract.TasksTable.COLUMN_IS_COMPLETED
import com.github.android.lvrn.lvrnproject.persistent.database.LavernaContract.TasksTable.COLUMN_NOTE_ID
import com.github.android.lvrn.lvrnproject.persistent.database.LavernaContract.TasksTable.TABLE_NAME
import com.github.android.lvrn.lvrnproject.persistent.repository.core.TaskRepository
import com.github.android.lvrn.lvrnproject.persistent.repository.impl.ProfileDependedRepositoryImpl
import com.github.android.lvrn.lvrnproject.util.PaginationArgs
import com.github.valhallalabs.laverna.persistent.entity.Task
import javax.inject.Inject

class TaskRepositoryImpl @Inject constructor() : ProfileDependedRepositoryImpl<Task>(TABLE_NAME), TaskRepository {

    override fun toContentValues(entity: Task): ContentValues {
        val completed = if (entity.isCompleted) 1 else 0
        return ContentValues().apply {
            put(LavernaBaseTable.COLUMN_ID, entity.id)
            put(ProfileDependedTable.COLUMN_PROFILE_ID, entity.profileId)
            put(COLUMN_NOTE_ID, entity.noteId)
            put(COLUMN_DESCRIPTION, entity.description)
            put(COLUMN_IS_COMPLETED, completed)
        }
    }

    override fun toEntity(cursor: Cursor): Task {
        return Task(
                cursor.getString(cursor.getColumnIndex(LavernaBaseTable.COLUMN_ID)),
                cursor.getString(cursor.getColumnIndex(ProfileDependedTable.COLUMN_PROFILE_ID)),
                cursor.getString(cursor.getColumnIndex(COLUMN_NOTE_ID)),
                cursor.getString(cursor.getColumnIndex(COLUMN_DESCRIPTION)),
                cursor.getInt(cursor.getColumnIndex(COLUMN_IS_COMPLETED)) > 0)
    }

    override fun getUncompletedByProfile(profileId: String, paginationArgs: PaginationArgs): MutableList<Task> {
        val additionalClause = " AND $COLUMN_IS_COMPLETED = 0"
        return super.getByIdCondition(ProfileDependedTable.COLUMN_PROFILE_ID, profileId, additionalClause, paginationArgs)
    }

    override fun getUncompletedByProfileAndDescription(profileId: String, description: String, paginationArgs: PaginationArgs): MutableList<Task> {
        val additionalClause = " AND $COLUMN_IS_COMPLETED = 0"
        return super.getByName(COLUMN_DESCRIPTION, profileId, description, additionalClause, paginationArgs)
    }

    override fun getByNote(noteId: String): MutableList<Task> {
        val query = ("SELECT * FROM " + TABLE_NAME
                + " WHERE " + COLUMN_NOTE_ID + " = '" + noteId + "'")
        return getByRawQuery(query)
    }

    override fun update(entity: Task): Boolean {
        val completed = if (entity.isCompleted) 1 else 0
        val query = ("UPDATE " + TABLE_NAME
                + " SET "
                + COLUMN_DESCRIPTION + "='" + entity.description + "', "
                + COLUMN_IS_COMPLETED + "='" + completed + "' "
                + " WHERE " + LavernaBaseTable.COLUMN_ID + "='" + entity.id + "'")
        return super.rawUpdateQuery(query)
    }
}